src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp

```diff
@@ -1237,6 +1244,234 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_SetupSrpHost(co
 }
 
+#if CHIP_DEVICE_CONFIG_ENABLE_THREAD_DNS_CLIENT
+template <class ImplClass>
+void GenericThreadStackManagerImpl_OpenThread<ImplClass>::FromOtDnsResponseToMdnsData(otDnsServiceInfo & serviceInfo,
+                                                                                      const char * serviceType,
+                                                                                      chip::Mdns::MdnsService & mdnsService,
+                                                                                      DnsServiceTxtEntries & serviceTxtEntries)
+{
+    char protocol[chip::Mdns::kMdnsProtocolTextMaxSize + 1];
+
+    // Extract from the <hostname>.<domain-name>. the <hostname> part.
+    size_t substringSize = strlen(serviceInfo.mHostNameBuffer) - strlen(strchr(serviceInfo.mHostNameBuffer, '.'));
+    strncpy(mdnsService.mHostName, serviceInfo.mHostNameBuffer, substringSize);
+    // Append string terminating character.
+    memset(mdnsService.mHostName + substringSize, '\0', 1);
+
+    // Extract from the <type>.<protocol>.<domain-name>. the <type> part.
+    substringSize = strlen(serviceType) - strlen(strchr(serviceType, '.'));
+    strncpy(mdnsService.mType, serviceType, substringSize);
+    // Append string terminating character.
+    memset(mdnsService.mType + substringSize, '\0', 1);
+
+    // Extract from the <type>.<protocol>.<domain-name>. the <protocol> part.
+    strncpy(protocol, serviceType + substringSize + 1, chip::Mdns::kMdnsProtocolTextMaxSize);
+    // Append string terminating character.
+    memset(protocol + chip::Mdns::kMdnsProtocolTextMaxSize, '\0', 1);
+
+    if (strncmp(protocol, "_udp", chip::Mdns::kMdnsProtocolTextMaxSize) == 0)
+    {
+        mdnsService.mProtocol = chip::Mdns::MdnsServiceProtocol::kMdnsProtocolUdp;
+    }
+    else if (strncmp(protocol, "_tcp", chip::Mdns::kMdnsProtocolTextMaxSize) == 0)
+    {
+        mdnsService.mProtocol = chip::Mdns::MdnsServiceProtocol::kMdnsProtocolTcp;
+    }
+    else
+    {
+        mdnsService.mProtocol = chip::Mdns::MdnsServiceProtocol::kMdnsProtocolUnknown;
+    }
+    mdnsService.mPort        = serviceInfo.mPort;
+    mdnsService.mInterface   = INET_NULL_INTERFACEID;
+    mdnsService.mAddressType = Inet::kIPAddressType_IPv6;
+    mdnsService.mAddress     = static_cast<chip::Optional<chip::Inet::IPAddress>>(ToIPAddress(serviceInfo.mHostAddress));
+
+    otDnsTxtEntryIterator iterator;
+    otDnsInitTxtEntryIterator(&iterator, serviceInfo.mTxtData, serviceInfo.mTxtDataSize);
+
+    otDnsTxtEntry txtEntry;
+
+    uint8_t entryIndex = 0;
+    while ((otDnsGetNextTxtEntry(&iterator, &txtEntry) == OT_ERROR_NONE) && entryIndex < kMaxDnsServiceTxtEntriesNumber)
+    {
+        if (txtEntry.mKey && strlen(txtEntry.mKey) < kMaxDnsServiceTxtKeySize)
+        {
+            strcpy(serviceTxtEntries.mTxtKeyBuffers[entryIndex], txtEntry.mKey);
+            serviceTxtEntries.mTxtEntries[entryIndex].mKey = &(serviceTxtEntries.mTxtKeyBuffers[entryIndex][0]);
```

Comment:
This can't be `serviceTxtEntries.mTxtEntries[entryIndex].mKey = serviceTxtEntries.mTxtKeyBuffers[entryIndex]`?